### ISSUE TITLE
Support carriageReturn to clear newline (e.g. loading bars)

### ISSUE THREAD

**Issue by clouless** [IonicaBizau/anser#18]
I am currently working on a Project, where I want to output shell-commands via a Websocket "live" to the browser. Your Project is amazing!

I was just wondering if you could support **ncurses** formatted output like complex loading bars.

I have this dummy-shell command I use to test the parsing:
https://gist.github.com/clouless/c60680a737be2b71c21d6e51cd8ec1b7

And that works very well with your tool except the ncurses part at the bottom.
Maybe that cannot be supported but I was thinking of some **Cursor-Reset** or **Screen-Reset** feature which could be included in the **toJSON** function.

Let's say we have a progress bar:

```
#!/bin/bash

echo "now comes the progress bar"
sleep 1s
echo -ne '#####                     (33%)\r'
sleep 1
echo -ne '#############             (66%)\r'
sleep 1
echo -ne '#######################   (100%)\r'
echo -ne '\n'
```

Maybe when that get's parsed the special chars that delete the line in the terminal lead to a new property inside the json like: `resetLastLine: false`

```
console.log(Anser.ansiToJson(txt));
// [ { content: 'now comes the progress bar',
//     fg: null,
//     bg: null,
//     fg_truecolor: null,
//     bg_truecolor: null,
//     was_processed: false,
//     resetLastLine: false,
//     isEmpty: [Function] },
// [ { content: '#####                     (33%)',
//     fg: null,
//     bg: null,
//     fg_truecolor: null,
//     bg_truecolor: null,
//     was_processed: false,
//     resetLastLine: true,
//     isEmpty: [Function] },
// [ { content: '#############             (66%)',
//     fg: null,
//     bg: null,
//     fg_truecolor: null,
//     bg_truecolor: null,
//     was_processed: false,
//     resetLastLine: true,
//     isEmpty: [Function] },
// [ { content: '#######################   (100%)',
//     fg: null,
//     bg: null,
//     fg_truecolor: null,
//     bg_truecolor: null,
//     was_processed: false,
//     resetLastLine: true,
//     isEmpty: [Function] },
```

When I then parse the JSON to use it in my custom Output in HTML I can interpret it and delete the last object. Basically I will have something like an ReactJS-Component for each Line of STDOUT in an Array and that get's rendered live. And once the `resetLastLine: true` comes in, I just remove the previous array element and  React Rerenders all components and removes the old ProgressBar and adds the new. This way it looks just like on the commandline - at least I hope so :D 

What do you think, could something like that be done?

**Comment by IonicaBizau** [comment]
That would be great! 

Since you use React, you may want to check this project: https://github.com/nteract/ansi-to-react
And a similar issue exists already here: https://github.com/nteract/ansi-to-react/issues/2

Contributions are welcome! :sparkle:

**Comment by clouless** [comment]
ok cool to hear :)

I was just using react as a reference, I am using http://riotjs.com/ but I want the implementation to be framework free. That is why I like the JSON thing.

Would you be willing to accept my Pull Request for the `resetLastLine` property when such a 'reset-line' character occurs in the output?

cheers.

**Comment by IonicaBizau** [comment]
@clouless Yes. Maybe we can make it generic at some point like: moving the cursor etc. `\r` does reset the line, but there are some other escape codes to move the cursor on a specific coordinate.

:+1:

**Comment by clouless** [comment]
ok cool. Maybe we can first write some kind of documentation about them? Because I don't know them at all. And then we can implement them step by step

**Comment by clouless** [comment]
I found this: http://www.tldp.org/HOWTO/Bash-Prompt-HOWTO/x361.html

And this would mean that the cursor can move freely, which could be a bit too much, but I will experiment with that for now. I will setup a proof of concept application to demonstrate the cursor-movement soon.

**Comment by IonicaBizau** [comment]
@clouless I guess that's a good one! 

Yes, better smaller/few changes / pull request. :grin:  It helps to ship new versions quickly.

**Comment by clouless** [comment]
Ok here is my PoC for the Shell Output:
https://github.com/clouless/poc-ansi-terminal-stdout-to-interactive-web

I will now try to create a Pull request to make the progress bar work :)

**Comment by clouless** [comment]
I just looked into the code and the main Problem I see is, that currently the json response is an array of chunks. 

```
line = [
  { object1 }
  { object2 }
]
```

Cooler would be 

```
line = {
 clearLine: true,
 chunks: [
    { object1 }
    { object2 }
  ]
}
```

So that we could set properties for a "line" because otherwise if I set `clearLine` on a chunk, I don't know how to handle that when using the info in an application. Because I don't know where the line starts.

**Comment by jerch** [comment]
@IonicaBizau For the escape codes you can have a look at vt100.net and the xterm docs http://invisible-island.net/xterm/ctlseqs/ctlseqs.html.

**Comment by clouless** [comment]
Ok it is now working for the **carriageReturn** and JSON:

![](https://camo.githubusercontent.com/63f1815e4806e16109496648efeed856419acdd7/68747470733a2f2f636c6f756c6573732e6769746875622e696f2f706f632d616e73692d7465726d696e616c2d7374646f75742d746f2d696e7465726163746976652d7765622f776562736f636b65742d636d642e6769663f7633)

SEE: https://github.com/clouless/poc-ansi-terminal-stdout-to-interactive-web

Pull Request: https://github.com/IonicaBizau/anser/pull/19

**Comment by clouless** [comment]
@IonicaBizau I am very confused by the code so my PR might not be the best solution. I would really like something like my suggestion above. To have a wrapper object for chunks. But for now that what my pr does would be enough for me.

**Comment by IonicaBizau** [comment]
> I am very confused by the code so my PR might not be the best solution.

Whoups!!!! Very sorry, looks like this is the transpiled version of the file. I just updated it: https://github.com/IonicaBizau/anser/blob/deba3fbe28decd3894810cfb5f3ff21c0531fa42/lib/index.js

Can you remake the changes and integrate them in this ES6 version?

**Comment by clouless** [comment]
Ok cool, Can you add an `.editorconfig` as well, because right know my editor fucks up all the intendentation. You can get it here: http://editorconfig.org/

**Comment by IonicaBizau** [comment]
@clouless Well, there's an issue about it too, didn't have time to work on it: https://github.com/IonicaBizau/code-style/issues/4

Contributions in that direction are more than welcome too!

<sub>PS: please, please without the *f* word––edited it  :grin:</sub>

**Comment by IonicaBizau** [comment]
Btw, that gif is amazing!

**Comment by clouless** [comment]
Ok cool. Can we just add a basic editorconfig then. Here is the PR:

https://github.com/IonicaBizau/anser/pull/20

Editorconfig is not supposed to do "linting" and "fixing" of lint-violations.
It is just about Tabs, indentation and line-endings. So I vote for merging the PR so that followup PRs do not mess up the diff.

**Comment by rgbkrk** [comment]
Cool happy to see carriage return handling here.

/cc @lgeiger

**Comment by clouless** [comment]
@IonicaBizau Ok. Here is my **Clean PR for carriageReturn Flag: #21**

**Comment by clouless** [comment]
@rgbkrk implemented while watching Falling Skies on netflix 😆

**Comment by IonicaBizau** [comment]
My scripts had a hard time to release a new version because somehow a `1.2.0` was already published as GitHub release.

@clouless's changes appear in both `1.2.0` and `1.2.1` npm versions, but the `1.2.0` git tag contains an older version (again, not sure).

https://ionicabizau.net/blog/33-how-i-npm

This problem should not appear for new releases. 😁

**Comment by IonicaBizau** [comment]
This explains why the transpiled version was pushed here previously—the automated process failed and remained into an inconsistent state. 💥

**Comment by clouless** [comment]
I think for me the carriageReturn feature is enough for now.
If anyone can suggest me a real use-case of something he wants to have done containing special chars, just ping me :) 

For now I am happy the feature got merged. Thanks a lot @IonicaBizau :)

**Comment by IonicaBizau** [comment]
@clouless Makes sense! I thank you, once again! 😁

@rgbkrk @jerch This may improve the `npm install` bars in the react thingy. We need to interpret that value to clear the line, I guess. 😄

**Comment by clouless** [comment]
I actually tried an `npm install` and parsed the result.
I just rendered it with monospaced font and replaced `\n` with `<br>` and that looked nice to me.

**Comment by rgbkrk** [comment]
@lgeiger ended up creating https://www.npmjs.com/package/escape-carriage to handle that for us. I'm happy with any amount of ansi -> json since this keeps us going down the path of being able to reduce from one "terminal state" to the next with more text. 😄

**Comment by lgeiger** [comment]
Big 👍 for carriage return handling.

**Comment by clouless** [comment]
@IonicaBizau can you please publish `1.2.1` to npmjs.com? Maybe you are already planning on doing so :) thx

**Comment by IonicaBizau** [comment]
@clouless I found what was actually wrong with this repository: using octals in strict mode. When trying to transpile, `babel-it` was complaining about it.

Addressed in #24.

:ship: 1.2.2 :tada: 

(for reference I published `1.2.1` too, but it's identical to `1.2.2.`)

Releasing next versions should be smoother. Haha! :joy:

**Comment by clouless** [comment]
ok thanks a lot 😄  I hope we will see times coming where we do not need babel anymore, should not be long ... 😆
